### PR TITLE
feat(api): api json command dispatcher and translator

### LINKED ISSUE #7428: Protocol Engine JSON Protocol translator
# Overview

The Protocol Engine is a rebuilding of how Protocols are executed.

Currently, JSON protocols are executed by translating JSON commands into `ProtocolContext` methods. 

This ticket creates a method to translate JSON protocol commands to a `ProtocolEngine` commands.

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

The basics of json protocol command dispatching with run-time models of JSON protocols and an interface to handle these commands. 

If we like this approach we can port the older JSON protocol execution using these classes.

closes #7428 

# Changelog

- created `protocols.runner.json_proto` package.
- Used `datamodel-code-generator` to generate pydantic models from the json schemas in shared data. See #7541 for technique used.
- Touch up generated models some to be have more usable naming and types.
- Created `AbstractCommandHandler` base class with methods to handle json protocol commands.
- Created `protocols.runner.json_proto.dispatcher` module with function `dispatch` method.  Function accepts a pydantic json protocol command model and a `AbstractCommandHandler`. 
- Added `CommandTranslator`. It is a `AbstractCommandHandler` that translates to protocol engine command requests. 

# Review requests

- Class / module naming.
- Package location/naming.
- Should we use generated pydantic models or stick with the typed dicts in shared data?

# Risk assessment

None

### CODE REVIEW

File: api/src/opentrons/protocols/runner/json_proto/command_translator.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+OnTranslatedCallback = Callable[[CommandRequestType], None]
+
+
+class CommandTranslator(AbstractCommandHandler):
+    """Class that translates commands from PD/JSON to ProtocolEngine."""
+
+    def __init__(
+            self,
+            state_view: StateView,
+            on_translated: OnTranslatedCallback) -> None:
```
Comment (mcous) [inline api/src/opentrons/protocols/runner/json_proto/command_translator.py]: At first glance, a callback interface to this class looks like it runs the risk of becoming awkward to use compared to a message signature that is:

```py
def translate(command: models.AllCommands) -> CommandRequestType:
```

What was the reasoning for going with the callback interface? Is there some need for inversion of control that I'm missing?

File: api/src/opentrons/protocols/runner/json_proto/models/json_protocol.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
+"""
```
Comment (mcous) [inline api/src/opentrons/protocols/runner/json_proto/models/json_protocol.py:1]: Probably worth a TODO comment to get this moved into shared-data eventually

**Review by mcous — COMMENTED** [review]
I think the Pydantic models are great. I'm a little confused by the callback interface to the translator, though, so I could use some additional info about what you're thinking

File: api/src/opentrons/protocols/runner/json_proto/command_translator.py (hunk 1 above)
Comment (amitlissack) [reply api/src/opentrons/protocols/runner/json_proto/command_translator.py]: This came from a discussion we had. Where there's no guarantee of 1-to-1 of JSON command to PE command. 

I could easily remove this. There are countless ways to implement this.

Comment (mcous) [reply api/src/opentrons/protocols/runner/json_proto/command_translator.py]: Ah that makes sense. Yeah I think I'm more partial to a collection return, but then again, I don't know how "flat maps" are typically accomplished in Python

Comment (amitlissack) [reply api/src/opentrons/protocols/runner/json_proto/command_translator.py]: I'll ditch it. I don't want future folks thinking there was too much thought put into it.

Comment (amitlissack) [reply api/src/opentrons/protocols/runner/json_proto/command_translator.py]: I look at this PR and wonder what the deuce I was thinking? Why an Abstract base class? Why a dispatcher? 

Back to the drawing board.

**Review by mcous — APPROVED** [review]

**Pull request merged**